tcn/tcn.py

```diff
@@ -121,30 +201,89 @@ def __init__(self,
             raise Exception()
 
-    def __call__(self, inputs, training=None):
-        x = inputs
-        # 1D FCN.
-        x = Conv1D(self.nb_filters, 1, padding=self.padding, kernel_initializer=self.kernel_initializer)(x)
-        skip_connections = []
+        # initialize parent class
+        super(TCN, self).__init__(**kwargs)
+
+    def build(self, input_shape):
+        self.main_conv1D = Conv1D(filters=self.nb_filters,
+                                  kernel_size=1,
+                                  padding=self.padding,
+                                  kernel_initializer=self.kernel_initializer)
+        self.main_conv1D.build(input_shape)
+
+        # member to hold current output shape of the layer for building purposes
+        self.build_output_shape = self.main_conv1D.compute_output_shape(input_shape)
+
+        # list to hold all the member ResidualBlocks
+        self.residual_blocks = list()
+
         for s in range(self.nb_stacks):
             for d in self.dilations:
-                x, skip_out = residual_block(x,
-                                             training=training,
-                                             dilation_rate=d,
-                                             nb_filters=self.nb_filters,
-                                             kernel_size=self.kernel_size,
-                                             padding=self.padding,
-                                             activation=self.activation,
-                                             dropout_rate=self.dropout_rate,
-                                             kernel_initializer=self.kernel_initializer,
-                                             use_batch_norm=self.use_batch_norm)
-                skip_connections.append(skip_out)
+                self.residual_blocks.append(ResidualBlock(dilation_rate=d,
+                                                          nb_filters=self.nb_filters,
+                                                          kernel_size=self.kernel_size,
+                                                          padding=self.padding,
+                                                          activation=self.activation,
+                                                          dropout_rate=self.dropout_rate,
+                                                          use_batch_norm=self.use_batch_norm,
+                                                          kernel_initializer=self.kernel_initializer,
+                                                          name='residual_block_{}'.format(len(self.residual_blocks))))
+                # build newest residual block
+                self.residual_blocks[-1].build(self.build_output_shape)
+                self.build_output_shape = self.residual_blocks[-1].res_output_shape
+
+        # this is done to force keras to add the layers in the list to self._layers
+        for layer in self.residual_blocks:
+            self.__setattr__(layer.name, layer)
```

Comment:
Are there any repercussions for having a slash in an attribute name? To my surprise, setattr with an invalid python identifier doesn't through an error so I guess it should be fine

Reply:
I did not even think about that, but there doesn't seem to be any.